lib/ret_web/controllers/api/v1/avatar_controller.ex

```diff
@@ -0,0 +1,192 @@
+      _ -> conn |> send_resp(404, "not found")
+    end
+  end
+
+  defp create_or_update(
+         conn,
+         _params,
+         %Avatar{account_id: avatar_account_id},
+         %Account{account_id: account_id}
+       )
+       when not is_nil(avatar_account_id) and avatar_account_id != account_id do
+    conn |> send_resp(401, "")
+  end
+
+  defp create_or_update(conn, params, avatar) do
+    account = conn |> Guardian.Plug.current_resource()
+
+    files_to_promotoe =
+      case params["files"] do
+        nil ->
+          %{}
+
+        files ->
+          files
+          |> Enum.map(fn {k, v} -> {String.to_atom(k), List.to_tuple(v)} end)
+          |> Enum.into(%{})
+      end
+
+    owned_file_results = Storage.promote(files_to_promotoe, account)
+
+    promotion_error =
+      owned_file_results |> Map.values() |> Enum.filter(&(elem(&1, 0) == :error)) |> Enum.at(0)
+
+    case promotion_error do
+      nil ->
+        owned_files =
+          owned_file_results |> Enum.map(fn {k, {:ok, file}} -> {k, file} end) |> Enum.into(%{})
+
+        parent_avatar =
+          if params["parent_avatar_id"] do
+            Repo.get_by(Avatar, avatar_sid: params["parent_avatar_id"])
+          end
+
+        {result, avatar} =
+          avatar
+          |> Avatar.changeset(account, owned_files, parent_avatar, params)
+          |> Repo.insert_or_update()
+
+        avatar = avatar |> Repo.preload(@file_columns)
+
+        case result do
+          :ok ->
+            conn |> render("create.json", avatar: avatar)
+
+          :error ->
+            conn |> send_resp(422, "invalid avatar")
+        end
+
+      {:error, :not_found} ->
+        conn |> send_resp(404, "no such file(s)")
+
+      {:error, :not_allowed} ->
+        conn |> send_resp(401, "")
+    end
+  end
+
+  defp collapse_avatar_files(%{parent_avatar: nil} = avatar),
+    do: avatar |> Map.take(@file_columns)
+
+  defp collapse_avatar_files(%{parent_avatar: parent} = avatar) do
+    Map.merge(collapse_avatar_files(parent), avatar |> Map.take(@file_columns), fn _k, v1, v2 ->
+      if is_nil(v2), do: v1, else: v2
+    end)
+  end
+
+  defp get_avatar(avatar_sid) do
+    avatar =
+      Avatar
+      |> Repo.get_by(avatar_sid: avatar_sid)
+      |> Repo.preload([@file_columns ++ [:parent_avatar, :account]])
+  end
+
+  def show(conn, %{"id" => avatar_sid}) do
+    conn |> show(avatar_sid |> get_avatar())
+  end
+
+  def show(conn, nil = _avatar) do
+    conn |> send_resp(404, "")
+  end
+
+  def show(conn, %Avatar{} = avatar) do
+    conn |> render("show.json", avatar: avatar)
+  end
+
+  def show_gltf(conn, %{"id" => avatar_sid}) do
+    conn |> show_gltf(Avatar |> Repo.get_by(avatar_sid: avatar_sid))
+  end
+
+  def show_gltf(conn, nil = _avatar) do
+    conn |> send_resp(404, "")
+  end
+
+  def show_gltf(conn, %Avatar{} = avatar) do
+    # TODO we ideally don't need to be featching the OwnedFiles until after we collapse them
+    avatar_files =
```

Comment:
i'd move this merging stuff out of the controller and probably into its own module `GLTFUtil` or `AvatarUtil` or something. also it'd be nice to have a simple positive unit test or two in place against the pure function parts just so as the semantics evolve over time we have the foundation to capture bugs and improvements under test.